Sources/Core/X86/HardwareBreakpointManager.cpp

```diff
@@ -218,4 +219,32 @@ ErrorCode HardwareBreakpointManager::isValid(Address const &address,
 
   return super::isValid(address, size, mode);
 }
+
+ErrorCode HardwareBreakpointManager::readDebugRegisters(
+    Target::Thread *thread, std::vector<uint64_t> &regs) const {
+  Architecture::CPUState state;
+
+  CHK(thread->readCPUState(state));
+
+#if defined(ARCH_X86)
+  for (int i = 0; i < kNumDebugRegisters; ++i) {
+    regs[i] = (i == 4 || i == 5) ? 0 : state.dr.dr[i];
+  }
+#elif defined(ARCH_X86_64)
+  for (int i = 0; i < kNumDebugRegisters; ++i) {
+    if (i < 4) {
+      regs[i] = state.is32 ? state.state32.dr.dr[i] : state.state64.dr.dr[i];
+    } else if (i == 4 || i == 5) {
+      regs[i] = 0;
+    } else {
+      regs[i] = state.is32 ? state.state32.dr.dr[i]
+                           : (state.state64.dr.dr[i] & 0xffffffff);
```

Comment:
I don't think you should worry about masking these here. The data available in the `CPUState` register should be already properly formatted (/masked) and you should read that directly so that all the clients can share that logic.

If you do that in CPUState, you can make this file simpler as well by having only one look and having an ifdef only for accessing `state32` and `state64`.